src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+                      :command command
+                      :version version
+                      :certname certname
+                      :received received
+                      :callback identity
+                      :annotations {:id (stock/entry-id entry)
+                                    :received received}})))
+
+(defn cmdref->cmd [q cmdref]
+  (let [entry (cmdref->entry cmdref)]
+    (with-open [command-stream (stock/stream q entry)]
+      (assoc cmdref
+             :payload (stream->json command-stream)
+             :entry entry))))
+
+(defn cons-attempt [cmdref exception]
+  (update cmdref :attempts conj {:exception exception
+                                 :time (kitchensink/timestamp)}))
+
+(defn store-command
+  ([q command version certname command-stream]
+   (store-command q command version certname command-stream identity))
+  ([q command version certname command-stream command-callback]
+   (let [current-time (time/now)
+         entry (stock/store q
+                            command-stream
+                            (metadata-str current-time command version certname))]
+     (map->CommandRef {:id (stock/entry-id entry)
+                       :command command
+                       :version version
+                       :certname certname
+                       :callback command-callback
+                       :received (kitchensink/timestamp current-time)
+                       :annotations {:id (stock/entry-id entry)
+                                     :received (kitchensink/timestamp current-time)}}))))
+
+(defn ack-command
+  [q command]
+  (stock/discard q (:entry command)))
+
+(deftype SortedCommandBuffer [^TreeMap fifo-queue ^HashMap certnames-map ^long max-entries ^clojure.lang.IFn delete-update-fn]
+  async-protos/Buffer
+  (full? [this]
+    (>= (.size fifo-queue) max-entries))
+
+  (remove! [this]
+    (let [^CommandRef cmdref (val (.pollFirstEntry fifo-queue))
+          command-type (:command cmdref)]
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (.remove certnames-map [command-type (:certname cmdref)]))
+      cmdref))
+
+  (add!* [this item]
+    (when-not (instance? CommandRef item)
+      (throw (IllegalArgumentException. (str "Cannot enqueue item of type " (class item)))))
+
+    (let [^CommandRef cmdref item
+          command-type (:command cmdref)
+          certname (:certname cmdref)]
+
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (when-let [^CommandRef old-command (.get certnames-map [command-type certname])]
+          (.put fifo-queue
+                (:id old-command)
+                (assoc old-command :delete? true))
+          (delete-update-fn (:command old-command) (:version old-command)))
+        (.put certnames-map [command-type certname] cmdref))
+      (.put fifo-queue (:id cmdref) cmdref))
```

Comment:
It may not matter if any exception in here is more globally fatal to pdb, but if not then I wonder if we might want to reorder or something, i.e. would we want to make sure the new command is in the queue before we mark the old one deleted?